Flagsmith.FlagsmithClient/FlagsmithClient.cs

```diff
@@ -0,0 +1,437 @@
+using FlagsmithEngine.Trait.Models;
+using Microsoft.Extensions.Logging;
+using System.Threading;
+using Flagsmith.Extensions;
+using System.Linq;
+
+namespace Flagsmith
+{   /// <summary>
+    /// A Flagsmith client.
+    /// Provides an interface for interacting with the Flagsmith http API.
+    /// </summary>
+    /// <exception cref="FlagsmithAPIError">
+    /// Thrown when error occurs during any http request to Flagsmith api.Not applicable for polling or ananlytics.
+    /// </exception>
+    /// <exception cref="FlagsmithClientError">
+    /// A general exception with a error message. Example: Feature not found, etc.
+    /// </exception>
+    public class FlagsmithClient
+    {
+        public static FlagsmithClient instance;
+
+        private readonly FlagsmithConfiguration configuration;
+        protected static HttpClient httpClient;
+        protected EnvironmentModel Environment { get; set; }
+        private readonly PollingManager _PollingManager;
+        private readonly IEngine _Engine;
+        private readonly AnalyticsProcessor _AnalyticsProcessor;
+
+        public FlagsmithClient(FlagsmithConfiguration flagsmithConfiguration)
+        {
+            if (flagsmithConfiguration == null)
+            {
+                throw new ArgumentNullException(nameof(flagsmithConfiguration),
+                    "Parameter must be provided when constructing an instance of the client.");
+            }
+
+            if (!flagsmithConfiguration.IsValid())
+            {
+                throw new ArgumentException("The provided configuration is not valid. An API Url and Environment Key must be provided.", nameof(flagsmithConfiguration));
+            }
+
+            if (instance == null)
+            {
+                configuration = flagsmithConfiguration;
+                var sp = ServicePointManager.FindServicePoint(new Uri(configuration.ApiUrl));
+                sp.ConnectionLeaseTimeout = 60 * 1000 * 5;
+                httpClient = new HttpClient();
+                instance = this;
+                _PollingManager = new PollingManager(GetAndUpdateEnvironmentFromApi, configuration.EnvironmentRefreshIntervalSeconds);
+                _Engine = new Engine();
+                if (configuration.EnableAnalytics)
+                    _AnalyticsProcessor = new AnalyticsProcessor(httpClient, configuration.EnvironmentKey, configuration.ApiUrl, configuration.Logger, configuration.CustomHeaders);
+                if (configuration.EnableClientSideEvaluation)
+                    _ = _PollingManager.StartPoll();
+
+            }
+            else
+            {
+                throw new NotSupportedException("FlagsmithClient should only be initialised once. Use FlagsmithClient.instance after successful initialisation");
+            }
+        }
+
+        /// <summary>
+        /// Get all the default for flags for the current environment.
+        /// </summary>
+        public async Task<List<Flag>> GetFeatureFlags()
```

Comment:
As per all our discussions, this class should only have 2 public methods (+ the one to update environment). Those methods should be named identically (barring snake / pascal case differences) to the ones in the python client. 

Please can you rewrite this client class so that it has only the methods available in the python client? Those methods should, as per the python client, return an object, from which you can access methods such as `IsFeatureEnabled` and `GetFeatureValue`. Those methods should not live on the client itself.